# Modern NLP in Python
### _- Or -_
## What you can learn about food by analyzing a million Yelp reviews

#### Before we get started...
__whois?__
- Patrick Harrison
- Lead Data Scientist @ S&P Global Market Intelligence - _**we are hiring**_
- University of Virginia &mdash; Systems Engineering
- patrick@skipgram.io / @skipgram

__Join Charlottesville Data Science!__
- On Meetup.com ... http://www.meetup.com/CharlottesvilleDataScience
- On Slack ... __https://cville.typeform.com/to/UEzMVh__
  - _link invites you to join the Cville team on Slack. Join Cville, then join channel __#datascience__._
  
_Note: I presented this notebook as a tutorial during the [PyData DC 2016 conference](http://pydata.org/dc2016/schedule/presentation/11/). To view the video of the presentation on YouTube, see [here](https://www.youtube.com/watch?v=6zm9NC9uRkk)._

## Our Trail Map
This tutorial features an end-to-end data science & natural language processing pipeline, starting with **raw data** and running through **preparing**, **modeling**, **visualizing**, and **analyzing** the data. We'll touch on the following points:
1. A tour of the dataset
1. Introduction to text processing with spaCy
1. Automatic phrase modeling
1. Topic modeling with LDA
1. Visualizing topic models with pyLDAvis
1. Word vector models with word2vec
1. Visualizing word2vec with t-SNE

...and we might even learn a thing or two about Python along the way.

Let's get started!

## The Yelp Dataset
[**The Yelp Dataset**](https://www.yelp.com/dataset_challenge/) is a dataset published by the business review service [Yelp](http://yelp.com) for academic research and educational purposes. I really like the Yelp dataset as a subject for machine learning and natural language processing demos, because it's big (but not so big that you need your own data center to process it), well-connected, and anyone can relate to it &mdash; it's largely about food, after all!

**Note:** If you'd like to execute this notebook interactively on your local machine, you'll need to download your own copy of the Yelp dataset. If you're reviewing a static copy of the notebook online, you can skip this step. Here's how to get the dataset:
1. Please visit the Yelp dataset webpage [here](https://www.yelp.com/dataset_challenge/)
1. Click "Get the Data"
1. Please review, agree to, and respect Yelp's terms of use!
1. The dataset downloads as a compressed .tgz file; uncompress it
1. Place the uncompressed dataset files (*yelp_academic_dataset_business.json*, etc.) in a directory named *yelp_dataset_challenge_academic_dataset*
1. Place the *yelp_dataset_challenge_academic_dataset* within the *data* directory in the *Modern NLP in Python* project folder

That's it! You're ready to go.

The current iteration of the Yelp dataset (as of this demo) consists of the following data:
- __552K__ users
- __77K__ businesses
- __2.2M__ user reviews

When focusing on restaurants alone, there are approximately __22K__ restaurants with approximately __1M__ user reviews written about them.

The data is provided in a handful of files in _.json_ format. We'll be using the following files for our demo:
- __yelp\_academic\_dataset\_business.json__ &mdash; _the records for individual businesses_
- __yelp\_academic\_dataset\_review.json__ &mdash; _the records for reviews users wrote about businesses_

The files are text files (UTF-8) with one _json object_ per line, each one corresponding to an individual data record. Let's take a look at a few examples.

In [1]:
import os
import codecs

data_directory = os.path.join('yelp_data')

businesses_filepath = os.path.join(data_directory,
                                   'yelp_academic_dataset_business.json')

with codecs.open(businesses_filepath, encoding='utf_8') as f:
    first_business_record = f.readline() 

print (first_business_record)

{"business_id":"0DI8Dt2PJp07XkVvIElIcQ","name":"Innovative Vapors","neighborhood":"","address":"227 E Baseline Rd, Ste J2","city":"Tempe","state":"AZ","postal_code":"85283","latitude":33.3782141,"longitude":-111.936102,"stars":4.5,"review_count":17,"is_open":0,"attributes":["BikeParking: True","BusinessAcceptsBitcoin: False","BusinessAcceptsCreditCards: True","BusinessParking: {'garage': False, 'street': False, 'validated': False, 'lot': True, 'valet': False}","DogsAllowed: False","RestaurantsPriceRange2: 2","WheelchairAccessible: True"],"categories":["Tobacco Shops","Nightlife","Vape Shops","Shopping"],"hours":["Monday 11:0-21:0","Tuesday 11:0-21:0","Wednesday 11:0-21:0","Thursday 11:0-21:0","Friday 11:0-22:0","Saturday 10:0-22:0","Sunday 11:0-18:0"],"type":"business"}



In [4]:
review_json_filepath = os.path.join(data_directory,
                                    'yelp_academic_dataset_review.json')

with codecs.open(review_json_filepath, encoding='utf_8') as f:
    first_review_record = f.readline()
    
print (first_review_record)

{"review_id":"NxL8SIC5yqOdnlXCg18IBg","user_id":"KpkOkG6RIf4Ra25Lhhxf1A","business_id":"2aFiy99vNLklCx3T_tGS9A","stars":5,"date":"2011-10-10","text":"If you enjoy service by someone who is as competent as he is personable, I would recommend Corey Kaplan highly. The time he has spent here has been very productive and working with him educational and enjoyable. I hope not to need him again (though this is highly unlikely) but knowing he is there if I do is very nice. By the way, I'm not from El Centro, CA. but Scottsdale, AZ.","useful":0,"funny":0,"cool":0,"type":"review"}



The business records consist of _key, value_ pairs containing information about the particular business. A few attributes we'll be interested in for this demo include:
- __business\_id__ &mdash; _unique identifier for businesses_
- __categories__ &mdash; _an array containing relevant category values of businesses_

The _categories_ attribute is of special interest. This demo will focus on restaurants, which are indicated by the presence of the _Restaurant_ tag in the _categories_ array. In addition, the _categories_ array may contain more detailed information about restaurants, such as the type of food they serve.

The review records are stored in a similar manner &mdash; _key, value_ pairs containing information about the reviews.

In [6]:
review_json_filepath = os.path.join(data_directory,
                                    'yelp_academic_dataset_review.json')

with codecs.open(review_json_filepath, encoding='utf_8') as f:
    first_review_record = f.readline()
    
print (first_review_record)

{"review_id":"NxL8SIC5yqOdnlXCg18IBg","user_id":"KpkOkG6RIf4Ra25Lhhxf1A","business_id":"2aFiy99vNLklCx3T_tGS9A","stars":5,"date":"2011-10-10","text":"If you enjoy service by someone who is as competent as he is personable, I would recommend Corey Kaplan highly. The time he has spent here has been very productive and working with him educational and enjoyable. I hope not to need him again (though this is highly unlikely) but knowing he is there if I do is very nice. By the way, I'm not from El Centro, CA. but Scottsdale, AZ.","useful":0,"funny":0,"cool":0,"type":"review"}



A few attributes of note on the review records:
- __business\_id__ &mdash; _indicates which business the review is about_
- __text__ &mdash; _the natural language text the user wrote_

The _text_ attribute will be our focus today!

_json_ is a handy file format for data interchange, but it's typically not the most usable for any sort of modeling work. Let's do a bit more data preparation to get our data in a more usable format. Our next code block will do the following:
1. Read in each business record and convert it to a Python `dict`
2. Filter out business records that aren't about restaurants (i.e., not in the "Restaurant" category)
3. Create a `frozenset` of the business IDs for restaurants, which we'll use in the next step

In [7]:
import json

restaurant_ids = set()

# open the businesses file
with codecs.open(businesses_filepath, encoding='utf_8') as f:
    
    # iterate through each line (json record) in the file
    for business_json in f:
        
        # convert the json record to a Python dict
        business = json.loads(business_json)
        
        # if this business is not a restaurant, skip to the next one
        try:
            if u'Restaurants' not in business[u'categories']:
                continue
        except:
            continue
            
        # add the restaurant business id to our restaurant_ids set
        restaurant_ids.add(business[u'business_id'])

# turn restaurant_ids into a frozenset, as we don't need to change it anymore
restaurant_ids = frozenset(restaurant_ids)

# print the number of unique restaurant ids in the dataset
print ('{:,}'.format(len(restaurant_ids)), u'restaurants in the dataset.')

48,485 restaurants in the dataset.


Next, we will create a new file that contains only the text from reviews about restaurants, with one review per line in the file.

In [9]:
intermediate_directory = os.path.join('intermediate')

review_txt_filepath = os.path.join(intermediate_directory,
                                   'review_text_all.txt')

In [43]:
%%time

# this is a bit time consuming - make the if statement True
# if you want to execute data prep yourself.
if 0 == 0:
    
    review_count = 0

    # create & open a new file in write mode
    with codecs.open(review_txt_filepath, 'w', encoding='utf_8') as review_txt_file:

        # open the existing review json file
        with codecs.open(review_json_filepath, encoding='utf_8') as review_json_file:

            # loop through all reviews in the existing file and convert to dict
            for review_json in review_json_file:
                review = json.loads(review_json)

                # if this review is not about a restaurant, skip to the next one
                if review[u'business_id'] not in restaurant_ids:
                    continue

                # write the restaurant review as a line in the new file
                # escape newline characters in the original review text
                review_txt_file.write(review[u'text'].replace('\n', '\\n') + '\n')
                review_count += 1
                if review_count>2000:
                    break

    print (u'''Text from {:,} restaurant reviews
              written to the new txt file.'''.format(review_count))
    
else:
    
    with codecs.open(review_txt_filepath, encoding='utf_8') as review_txt_file:
        for review_count, line in enumerate(review_txt_file):
            pass
        
    print (u'Text from {:,} restaurant reviews in the txt file.'.format(review_count + 1))

Text from 2,001 restaurant reviews
              written to the new txt file.
CPU times: user 212 ms, sys: 196 ms, total: 408 ms
Wall time: 412 ms


## spaCy &mdash; Industrial-Strength NLP in Python

![spaCy](https://s3.amazonaws.com/skipgram-images/spaCy.png)

[**spaCy**](https://spacy.io) is an industrial-strength natural language processing (_NLP_) library for Python. spaCy's goal is to take recent advancements in natural language processing out of research papers and put them in the hands of users to build production software.

spaCy handles many tasks commonly associated with building an end-to-end natural language processing pipeline:
- Tokenization
- Text normalization, such as lowercasing, stemming/lemmatization
- Part-of-speech tagging
- Syntactic dependency parsing
- Sentence boundary detection
- Named entity recognition and annotation

In the "batteries included" Python tradition, spaCy contains built-in data and models which you can use out-of-the-box for processing general-purpose English language text:
- Large English vocabulary, including stopword lists
- Token "probabilities"
- Word vectors

spaCy is written in optimized Cython, which means it's _fast_. According to a few independent sources, it's the fastest syntactic parser available in any language. Key pieces of the spaCy parsing pipeline are written in pure C, enabling efficient multithreading (i.e., spaCy can release the _GIL_).

In [44]:
import spacy
import pandas as pd
import itertools as it

nlp = spacy.load('en')

Let's grab a sample review to play with.

In [45]:
with codecs.open(review_txt_filepath, encoding='utf_8') as f:
    sample_review = list(it.islice(f, 10, 11))[0]
    sample_review = sample_review.replace('\\n', '\n')
        
print (sample_review)

Okay, so it is a bar with an old fashioned family restaurant attached. We were there almost 2 hours for the Tuesday night burger special. They need more staff. The table was a little sticky. However, the food was delicious. The French fries were nice and crispy, and the Half pound burger didn't taste like a pre-made frozen patty. Our total bill was $15.30 for 2 meals and 2 Cokes. We'll definitely go back.



Hand the review text to spaCy, and be prepared to wait...

In [46]:
%%time
parsed_review = nlp(sample_review)

CPU times: user 4 ms, sys: 4 ms, total: 8 ms
Wall time: 11.8 ms


...1/20th of a second or so. Let's take a look at what we got during that time...

In [47]:
print (parsed_review)

Okay, so it is a bar with an old fashioned family restaurant attached. We were there almost 2 hours for the Tuesday night burger special. They need more staff. The table was a little sticky. However, the food was delicious. The French fries were nice and crispy, and the Half pound burger didn't taste like a pre-made frozen patty. Our total bill was $15.30 for 2 meals and 2 Cokes. We'll definitely go back.



Looks the same! What happened under the hood?

What about sentence detection and segmentation?

In [48]:
for num, sentence in enumerate(parsed_review.sents):
    print ('Sentence {}:'.format(num + 1))
    print (sentence)
    print ('')

Sentence 1:
Okay, so it is a bar with an old fashioned family restaurant attached.

Sentence 2:
We were there almost 2 hours for the Tuesday night burger special.

Sentence 3:
They need more staff.

Sentence 4:
The table was a little sticky.

Sentence 5:
However, the food was delicious.

Sentence 6:
The French fries were nice and crispy, and the Half pound burger didn't taste like a pre-made frozen patty.

Sentence 7:
Our total bill was $15.30 for 2 meals and 2 Cokes.

Sentence 8:
We'll definitely go back.




What about named entity detection?

In [49]:
for num, entity in enumerate(parsed_review.ents):
    print ('Entity {}:'.format(num + 1), entity, '-', entity.label_)
    print ('')

Entity 1: almost 2 hours - TIME

Entity 2: Tuesday - DATE

Entity 3: night - TIME

Entity 4: French - NORP

Entity 5: Half - CARDINAL

Entity 6: 15.30 - MONEY

Entity 7: 2 - CARDINAL

Entity 8: 2 - CARDINAL



What about part of speech tagging?

In [50]:
token_text = [token.orth_ for token in parsed_review]
token_pos = [token.pos_ for token in parsed_review]

pd.DataFrame({'token_text':token_text,'part_of_speech':token_pos})

part_of_speech  token_text
0            INTJ        Okay
1           PUNCT           ,
2           CCONJ          so
3            PRON          it
4            VERB          is
5             DET           a
6            NOUN         bar
7             ADP        with
8             DET          an
9             ADJ         old
10           VERB   fashioned
11           NOUN      family
12           NOUN  restaurant
13           VERB    attached
14          PUNCT           .
15           PRON          We
16           VERB        were
17            ADV       there
18            ADV      almost
19            NUM           2
20           NOUN       hours
21            ADP         for
22            DET         the
23          PROPN     Tuesday
24           NOUN       night
25           NOUN      burger
26            ADJ     special
27          PUNCT           .
28           PRON        They
29           VERB        need
..            ...         ...
61            ADV         n't
62           VERB       taste
63            ADP        like
64            DET           a
65           NOUN         pre
66          PUNCT           -
67           VERB        made
68            ADJ      frozen
69           NOUN       patty
70          PUNCT           .
71            ADJ         Our
72            ADJ       total
73           NOUN        bill
74           VERB         was
75            SYM           $
76            NUM       15.30
77            ADP         for
78            NUM           2
79           NOUN       meals
80          CCONJ         and
81            NUM           2
82           NOUN       Cokes
83          PUNCT           .
84           PRON          We
85           VERB         'll
86            ADV  definitely
87           VERB          go
88            ADV        back
89          PUNCT           .
90          SPACE          \n

[91 rows x 2 columns]

What about text normalization, like stemming/lemmatization and shape analysis?

In [51]:
token_lemma = [token.lemma_ for token in parsed_review]
token_shape = [token.shape_ for token in parsed_review]

pd.DataFrame({'token_text':token_text,'token_lemma':token_lemma,'token_shape':token_shape})

token_lemma token_shape  token_text
0         okay        Xxxx        Okay
1            ,           ,           ,
2           so          xx          so
3       -PRON-          xx          it
4           be          xx          is
5            a           x           a
6          bar         xxx         bar
7         with        xxxx        with
8           an          xx          an
9          old         xxx         old
10     fashion        xxxx   fashioned
11      family        xxxx      family
12  restaurant        xxxx  restaurant
13      attach        xxxx    attached
14           .           .           .
15      -PRON-          Xx          We
16          be        xxxx        were
17       there        xxxx       there
18      almost        xxxx      almost
19           2           d           2
20        hour        xxxx       hours
21         for         xxx         for
22         the         xxx         the
23     tuesday       Xxxxx     Tuesday
24       night        xxxx       night
25      burger        xxxx      burger
26     special        xxxx     special
27           .           .           .
28      -PRON-        Xxxx        They
29          ne        xxxx        need
..         ...         ...         ...
61         not         x'x         n't
62       taste        xxxx       taste
63        like        xxxx        like
64           a           x           a
65         pre         xxx         pre
66           -           -           -
67        make        xxxx        made
68      frozen        xxxx      frozen
69       patty        xxxx       patty
70           .           .           .
71      -PRON-         Xxx         Our
72       total        xxxx       total
73        bill        xxxx        bill
74          be         xxx         was
75           $           $           $
76       15.30       dd.dd       15.30
77         for         xxx         for
78           2           d           2
79        meal        xxxx       meals
80         and         xxx         and
81           2           d           2
82        coke       Xxxxx       Cokes
83           .           .           .
84      -PRON-          Xx          We
85        will         'xx         'll
86  definitely        xxxx  definitely
87          go          xx          go
88        back        xxxx        back
89           .           .           .
90          \n          \n          \n

[91 rows x 3 columns]

What about token-level entity analysis?

In [52]:
token_entity_type = [token.ent_type_ for token in parsed_review]
token_entity_iob = [token.ent_iob_ for token in parsed_review]

pd.DataFrame({'token_text':token_text,'entity_type':token_entity_type,'inside_outside_begin':token_entity_iob})

entity_type inside_outside_begin  token_text
0                                 O        Okay
1                                 O           ,
2                                 O          so
3                                 O          it
4                                 O          is
5                                 O           a
6                                 O         bar
7                                 O        with
8                                 O          an
9                                 O         old
10                                O   fashioned
11                                O      family
12                                O  restaurant
13                                O    attached
14                                O           .
15                                O          We
16                                O        were
17                                O       there
18        TIME                    B      almost
19        TIME                    I           2
20        TIME                    I       hours
21                                O         for
22                                O         the
23        DATE                    B     Tuesday
24        TIME                    B       night
25                                O      burger
26                                O     special
27                                O           .
28                                O        They
29                                O        need
..         ...                  ...         ...
61                                O         n't
62                                O       taste
63                                O        like
64                                O           a
65                                O         pre
66                                O           -
67                                O        made
68                                O      frozen
69                                O       patty
70                                O           .
71                                O         Our
72                                O       total
73                                O        bill
74                                O         was
75                                O           $
76       MONEY                    B       15.30
77                                O         for
78    CARDINAL                    B           2
79                                O       meals
80                                O         and
81    CARDINAL                    B           2
82                                O       Cokes
83                                O           .
84                                O          We
85                                O         'll
86                                O  definitely
87                                O          go
88                                O        back
89                                O           .
90                                O          \n

[91 rows x 3 columns]

What about a variety of other token-level attributes, such as the relative frequency of tokens, and whether or not a token matches any of these categories?
- stopword
- punctuation
- whitespace
- represents a number
- whether or not the token is included in spaCy's default vocabulary?

In [53]:
token_attributes = [(token.orth_,
                     token.prob,
                     token.is_stop,
                     token.is_punct,
                     token.is_space,
                     token.like_num,
                     token.is_oov)
                    for token in parsed_review]

df = pd.DataFrame(token_attributes,
                  columns=['text',
                           'log_probability',
                           'stop?',
                           'punctuation?',
                           'whitespace?',
                           'number?',
                           'out of vocab.?'])

df.loc[:, 'stop?':'out of vocab.?'] = (df.loc[:, 'stop?':'out of vocab.?']
                                       .applymap(lambda x: u'Yes' if x else u''))
                                               
df

text  log_probability stop? punctuation? whitespace? number?  \
0         Okay       -10.330239                                          
1            ,        -3.391480                Yes                       
2           so        -6.064945   Yes                                    
3           it        -4.506450   Yes                                    
4           is        -4.329765   Yes                                    
5            a        -3.983075   Yes                                    
6          bar        -9.947871                                          
7         with        -5.363765   Yes                                    
8           an        -5.953294   Yes                                    
9          old        -7.795482                                          
10   fashioned       -19.579313                                          
11      family        -8.719374                                          
12  restaurant       -11.034676                                          
13    attached       -10.961068                                          
14           .        -3.072948                Yes                       
15          We        -7.376642   Yes                                    
16        were        -6.634815   Yes                                    
17       there        -6.153741   Yes                                    
18      almost        -8.373269   Yes                                    
19           2        -7.786293                                    Yes   
20       hours        -8.905840                                          
21         for        -4.913970   Yes                                    
22         the        -3.425446   Yes                                    
23     Tuesday       -11.236149                                          
24       night        -8.964851                                          
25      burger       -19.579313                                          
26     special        -9.451114                                          
27           .        -3.072948                Yes                       
28        They        -6.996389   Yes                                    
29        need        -7.108060                                          
..         ...              ...   ...          ...         ...     ...   
61         n't        -4.883962                                          
62       taste       -10.099010                                          
63        like        -5.808349                                          
64           a        -3.983075   Yes                                    
65         pre        -9.939320                                          
66           -        -5.202416                Yes                       
67        made        -7.493931   Yes                                    
68      frozen       -11.543664                                          
69       patty       -19.579313                                          
70           .        -3.072948                Yes                       
71         Our        -9.764631   Yes                                    
72       total        -9.346201                                          
73        bill        -9.207648                                          
74         was        -5.404201   Yes                                    
75           $        -7.211836                                          
76       15.30       -19.579313                                    Yes   
77         for        -4.913970   Yes                                    
78           2        -7.786293                                    Yes   
79       meals       -19.579313                                          
80         and        -4.195279   Yes                                    
81           2        -7.786293                                    Yes   
82       Cokes       -19.579313                                          
83           .

If the text you'd like to process is general-purpose English language text (i.e., not domain-specific, like medical literature), spaCy is ready to use out-of-the-box.

I think it will eventually become a core part of the Python data science ecosystem &mdash; it will do for natural language computing what other great libraries have done for numerical computing.

## Phrase Modeling

_Phrase modeling_ is another approach to learning combinations of tokens that together represent meaningful multi-word concepts. We can develop phrase models by looping over the the words in our reviews and looking for words that _co-occur_ (i.e., appear one after another) together much more frequently than you would expect them to by random chance. The formula our phrase models will use to determine whether two tokens $A$ and $B$ constitute a phrase is:

$$\frac{count(A\ B) - count_{min}}{count(A) * count(B)} * N > threshold$$

...where:
* $count(A)$ is the number of times token $A$ appears in the corpus
* $count(B)$ is the number of times token $B$ appears in the corpus
* $count(A\ B)$ is the number of times the tokens $A\ B$ appear in the corpus *in order*
* $N$ is the total size of the corpus vocabulary
* $count_{min}$ is a user-defined parameter to ensure that accepted phrases occur a minimum number of times
* $threshold$ is a user-defined parameter to control how strong of a relationship between two tokens the model requires before accepting them as a phrase

Once our phrase model has been trained on our corpus, we can apply it to new text. When our model encounters two tokens in new text that identifies as a phrase, it will merge the two into a single new token.

Phrase modeling is superficially similar to named entity detection in that you would expect named entities to become phrases in the model (so _new york_ would become *new\_york*). But you would also expect multi-word expressions that represent common concepts, but aren't specifically named entities (such as _happy hour_) to also become phrases in the model.

We turn to the indispensible [**gensim**](https://radimrehurek.com/gensim/index.html) library to help us with phrase modeling &mdash; the [**Phrases**](https://radimrehurek.com/gensim/models/phrases.html) class in particular.

In [54]:
from gensim.models import Phrases
from gensim.models.word2vec import LineSentence

As we're performing phrase modeling, we'll be doing some iterative data transformation at the same time. Our roadmap for data preparation includes:

1. Segment text of complete reviews into sentences & normalize text
1. First-order phrase modeling $\rightarrow$ _apply first-order phrase model to transform sentences_
1. Second-order phrase modeling $\rightarrow$ _apply second-order phrase model to transform sentences_
1. Apply text normalization and second-order phrase model to text of complete reviews

We'll use this transformed data as the input for some higher-level modeling approaches in the following sections.

First, let's define a few helper functions that we'll use for text normalization. In particular, the `lemmatized_sentence_corpus` generator function will use spaCy to:
- Iterate over the 1M reviews in the `review_txt_all.txt` we created before
- Segment the reviews into individual sentences
- Remove punctuation and excess whitespace
- Lemmatize the text

... and do so efficiently in parallel, thanks to spaCy's `nlp.pipe()` function.

In [55]:
def punct_space(token):
    """
    helper function to eliminate tokens
    that are pure punctuation or whitespace
    """
    
    return token.is_punct or token.is_space

def line_review(filename):
    """
    generator function to read in reviews from the file
    and un-escape the original line breaks in the text
    """
    
    with codecs.open(filename, encoding='utf_8') as f:
        for review in f:
            yield review.replace('\\n', '\n')
            
def lemmatized_sentence_corpus(filename):
    """
    generator function to use spaCy to parse reviews,
    lemmatize the text, and yield sentences
    """
    
    for parsed_review in nlp.pipe(line_review(filename),
                                  batch_size=10000, n_threads=4):
        
        for sent in parsed_review.sents:
            yield u' '.join([token.lemma_ for token in sent
                             if not punct_space(token)])

In [56]:
unigram_sentences_filepath = os.path.join(intermediate_directory,
                                          'unigram_sentences_all.txt')

Let's use the `lemmatized_sentence_corpus` generator to loop over the original review text, segmenting the reviews into individual sentences and normalizing the text. We'll write this data back out to a new file (`unigram_sentences_all`), with one normalized sentence per line. We'll use this data for learning our phrase models.

In [57]:
%%time

# this is a bit time consuming - make the if statement True
# if you want to execute data prep yourself.
if 0 == 0:

    with codecs.open(unigram_sentences_filepath, 'w', encoding='utf_8') as f:
        for sentence in lemmatized_sentence_corpus(review_txt_filepath):
            f.write(sentence + '\n')

CPU times: user 21.1 s, sys: 36 ms, total: 21.1 s
Wall time: 9.37 s


If your data is organized like our `unigram_sentences_all` file now is &mdash; a large text file with one document/sentence per line &mdash; gensim's [**LineSentence**](https://radimrehurek.com/gensim/models/word2vec.html#gensim.models.word2vec.LineSentence) class provides a convenient iterator for working with other gensim components. It *streams* the documents/sentences from disk, so that you never have to hold the entire corpus in RAM at once. This allows you to scale your modeling pipeline up to potentially very large corpora.

In [41]:
unigram_sentences = LineSentence(unigram_sentences_filepath)

Let's take a look at a few sample sentences in our new, transformed file.

In [58]:
for unigram_sentence in it.islice(unigram_sentences, 230, 240):
    print (u' '.join(unigram_sentence))
    print (u'')

-PRON- order the salt and pepper prawns and the waiter express to -PRON- that -PRON- have a few shell underneath however the prawns come whole fry with shell and eye

-PRON- mention that -PRON- be expect a few shell but could not eat through these,,,he say well -PRON- tell -PRON- that -PRON- come with shell and half heartely offer to recook -PRON-

this only after the plate have be sit for twenty minute

-PRON- avoid -PRON- table like the plague as -PRON- could tell -PRON- could clearly see -PRON- be not eat

at the end -PRON- offer a free dessert but by that time -PRON- be a little too late for -PRON-

-PRON- do not like to compare high end asian restaurant to pf changs but on this night pf would win

try the grand wok in the mgm grand

only 4 star because of the expense without -PRON- $ 55 groupon -PRON- meal for two include one large house sake would have be over $ 100

however even if -PRON- have not have the groupon -PRON- would have be well worth the money

what a find



Next, we'll learn a phrase model that will link individual words into two-word phrases. We'd expect words that together represent a specific concept, like "`ice cream`", to be linked together to form a new, single token: "`ice_cream`".

In [59]:
bigram_model_filepath = os.path.join(intermediate_directory, 'bigram_model_all')

In [60]:
%%time

# this is a bit time consuming - make the if statement True
# if you want to execute modeling yourself.
if 0 == 0:

    bigram_model = Phrases(unigram_sentences)

    bigram_model.save(bigram_model_filepath)
    
# load the finished model from disk
bigram_model = Phrases.load(bigram_model_filepath)

CPU times: user 760 ms, sys: 28 ms, total: 788 ms
Wall time: 775 ms


Now that we have a trained phrase model for word pairs, let's apply it to the review sentences data and explore the results.

In [67]:
bigram_sentences_filepath = os.path.join(intermediate_directory,
                                         'bigram_sentences_all.txt')

In [68]:
%%time

# this is a bit time consuming - make the if statement True
# if you want to execute data prep yourself.
if 0 == 0:

    with codecs.open(bigram_sentences_filepath, 'w', encoding='utf_8') as f:
        
        for unigram_sentence in unigram_sentences:
            
            bigram_sentence = u' '.join(bigram_model[unigram_sentence])
            
            f.write(bigram_sentence + '\n')

/usr/local/lib/python3.5/dist-packages/gensim/models/phrases.py:274: UserWarning: For a faster implementation, use the gensim.models.phrases.Phraser class
  warnings.warn("For a faster implementation, use the gensim.models.phrases.Phraser class")


CPU times: user 1.42 s, sys: 0 ns, total: 1.42 s
Wall time: 1.41 s


In [69]:
bigram_sentences = LineSentence(bigram_sentences_filepath)

In [70]:
for bigram_sentence in it.islice(bigram_sentences, 230, 240):
    print (u' '.join(bigram_sentence))
    print (u'')

-PRON- order the salt and pepper prawns and the waiter express to -PRON- that -PRON- have a_few shell underneath however the prawns come whole fry with shell and eye

-PRON- mention that -PRON- be expect a_few shell but could not eat through these,,,he say well -PRON- tell -PRON- that -PRON- come with shell and half heartely offer to recook -PRON-

this only after the plate have be sit for twenty minute

-PRON- avoid -PRON- table like the plague as -PRON- could tell -PRON- could clearly see -PRON- be not eat

at the end -PRON- offer a free dessert but by that time -PRON- be a_little too late for -PRON-

-PRON- do_not like to compare high_end asian restaurant to pf changs but on this night pf would win

try the grand wok in the mgm grand

only 4_star because of the expense without -PRON- $ 55 groupon -PRON- meal for two include one large house sake would have be over $_100

however even if -PRON- have not have the groupon -PRON- would have be well_worth the money

what a find



Looks like the phrase modeling worked! We now see two-word phrases, such as "`ice_cream`" and "`apple_pie`", linked together in the text as a single token. Next, we'll train a _second-order_ phrase model. We'll apply the second-order phrase model on top of the already-transformed data, so that incomplete word combinations like "`vanilla_ice cream`" will become fully joined to "`vanilla_ice_cream`". No disrespect intended to [Vanilla Ice](https://www.youtube.com/watch?v=rog8ou-ZepE), of course.

In [71]:
trigram_model_filepath = os.path.join(intermediate_directory,
                                      'trigram_model_all')

In [72]:
%%time

# this is a bit time consuming - make the if statement True
# if you want to execute modeling yourself.
if 0 == 0:

    trigram_model = Phrases(bigram_sentences)

    trigram_model.save(trigram_model_filepath)
    
# load the finished model from disk
trigram_model = Phrases.load(trigram_model_filepath)

CPU times: user 768 ms, sys: 20 ms, total: 788 ms
Wall time: 775 ms


We'll apply our trained second-order phrase model to our first-order transformed sentences, write the results out to a new file, and explore a few of the second-order transformed sentences.

In [73]:
trigram_sentences_filepath = os.path.join(intermediate_directory,
                                          'trigram_sentences_all.txt')

In [77]:
%%time

# this is a bit time consuming - make the if statement True
# if you want to execute data prep yourself.
if 0 == 0:

    with codecs.open(trigram_sentences_filepath, 'w', encoding='utf_8') as f:
        
        for bigram_sentence in bigram_sentences:
            
            trigram_sentence = u' '.join(trigram_model[bigram_sentence])
            
            f.write(trigram_sentence + '\n')

/usr/local/lib/python3.5/dist-packages/gensim/models/phrases.py:274: UserWarning: For a faster implementation, use the gensim.models.phrases.Phraser class
  warnings.warn("For a faster implementation, use the gensim.models.phrases.Phraser class")


CPU times: user 1.38 s, sys: 8 ms, total: 1.39 s
Wall time: 1.39 s


In [78]:
trigram_sentences = LineSentence(trigram_sentences_filepath)

In [79]:
for trigram_sentence in it.islice(trigram_sentences, 230, 240):
    print (u' '.join(trigram_sentence))
    print (u'')

-PRON- order the salt and pepper prawns and the waiter express to -PRON- that -PRON- have a_few shell underneath however the prawns come whole fry with shell and eye

-PRON- mention that -PRON- be expect a_few shell but could_not eat through these,,,he say well -PRON- tell -PRON- that -PRON- come_with shell and half heartely offer to recook -PRON-

this only after the plate have be sit for twenty minute

-PRON- avoid -PRON- table like the plague as -PRON- could tell -PRON- could clearly see -PRON- be not eat

at the end -PRON- offer a free dessert but by that time -PRON- be a_little too late for -PRON-

-PRON- do_not like to compare high_end asian restaurant to pf changs but on this night pf would win

try the grand wok in the mgm grand

only 4_star because of the expense without -PRON- $ 55 groupon -PRON- meal for two include one large house sake would have be over $_100

however even if -PRON- have not have the groupon -PRON- would have be well_worth the money

what a find



Looks like the second-order phrase model was successful. We're now seeing three-word phrases, such as "`vanilla_ice_cream`" and "`cinnamon_ice_cream`".

The final step of our text preparation process circles back to the complete text of the reviews. We're going to run the complete text of the reviews through a pipeline that applies our text normalization and phrase models.

In addition, we'll remove stopwords at this point. _Stopwords_ are very common words, like _a_, _the_, _and_, and so on, that serve functional roles in natural language, but typically don't contribute to the overall meaning of text. Filtering stopwords is a common procedure that allows higher-level NLP modeling techniques to focus on the words that carry more semantic weight.

Finally, we'll write the transformed text out to a new file, with one review per line.

In [80]:
trigram_reviews_filepath = os.path.join(intermediate_directory,
                                        'trigram_transformed_reviews_all.txt')

In [82]:
%%time

# this is a bit time consuming - make the if statement True
# if you want to execute data prep yourself.
if 0 == 0:

    with codecs.open(trigram_reviews_filepath, 'w', encoding='utf_8') as f:
        
        for parsed_review in nlp.pipe(line_review(review_txt_filepath),
                                      batch_size=10000, n_threads=4):
            
            # lemmatize the text, removing punctuation and whitespace
            unigram_review = [token.lemma_ for token in parsed_review
                              if not punct_space(token)]
            
            # apply the first-order and second-order phrase models
            bigram_review = bigram_model[unigram_review]
            trigram_review = trigram_model[bigram_review]
            
            # remove any remaining stopwords
            trigram_review = [term for term in trigram_review
                              if term not in spacy.en.language_data.STOP_WORDS]
            
            # write the transformed review as a line in the new file
            trigram_review = u' '.join(trigram_review)
            f.write(trigram_review + '\n')

/usr/local/lib/python3.5/dist-packages/gensim/models/phrases.py:274: UserWarning: For a faster implementation, use the gensim.models.phrases.Phraser class
  warnings.warn("For a faster implementation, use the gensim.models.phrases.Phraser class")


CPU times: user 22.6 s, sys: 44 ms, total: 22.7 s
Wall time: 11 s


Let's preview the results. We'll grab one review from the file with the original, untransformed text, grab the same review from the file with the normalized and transformed text, and compare the two.

In [84]:
print (u'Original:' + u'\n')

for review in it.islice(line_review(review_txt_filepath), 11, 12):
    print (review)

print (u'----' + u'\n')
print (u'Transformed:' + u'\n')

with codecs.open(trigram_reviews_filepath, encoding='utf_8') as f:
    for review in it.islice(f, 11, 12):
        print (review)

Original:

3 stars is fair for this pub. 

The food is good and well priced for a bar. 

Tuesday has $5 burgers and chicken sandwiches. 

The pecan pie is good for dessert. 

I got the chicken dinner, it came with a 6 oz chicken, broccoli, steamed potatoes, salad, and roll. It was ok. I'd stick with the burgers though. The salad was fresh. 

Their fries are really good.

----

Transformed:

3_star fair pub food good price bar tuesday $_5 burger chicken sandwich pecan pie good dessert -PRON- chicken dinner -PRON- come_with 6 oz chicken broccoli steam potato_salad roll -PRON- ok -PRON- stick burger salad fresh -PRON- fry good



You can see that most of the grammatical structure has been scrubbed from the text &mdash; capitalization, articles/conjunctions, punctuation, spacing, etc. However, much of the general semantic *meaning* is still present. Also, multi-word concepts such as "`friday_night`" and "`above_average`" have been joined into single tokens, as expected. The review text is now ready for higher-level modeling. 

## Topic Modeling with Latent Dirichlet Allocation (_LDA_)

*Topic modeling* is family of techniques that can be used to describe and summarize the documents in a corpus according to a set of latent "topics". For this demo, we'll be using [*Latent Dirichlet Allocation*](http://www.jmlr.org/papers/volume3/blei03a/blei03a.pdf) or LDA, a popular approach to topic modeling.

In many conventional NLP applications, documents are represented a mixture of the individual tokens (words and phrases) they contain. In other words, a document is represented as a *vector* of token counts. There are two layers in this model &mdash; documents and tokens &mdash; and the size or dimensionality of the document vectors is the number of tokens in the corpus vocabulary. This approach has a number of disadvantages:
* Document vectors tend to be large (one dimension for each token $\Rightarrow$ lots of dimensions)
* They also tend to be very sparse. Any given document only contains a small fraction of all tokens in the vocabulary, so most values in the document's token vector are 0.
* The dimensions are fully indepedent from each other &mdash; there's no sense of connection between related tokens, such as _knife_ and _fork_.

LDA injects a third layer into this conceptual model. Documents are represented as a mixture of a pre-defined number of *topics*, and the *topics* are represented as a mixture of the individual tokens in the vocabulary. The number of topics is a model hyperparameter selected by the practitioner. LDA makes a prior assumption that the (document, topic) and (topic, token) mixtures follow [*Dirichlet*](https://en.wikipedia.org/wiki/Dirichlet_distribution) probability distributions. This assumption encourages documents to consist mostly of a handful of topics, and topics to consist mostly of a modest set of the tokens.

![LDA](https://s3.amazonaws.com/skipgram-images/LDA.png)

LDA is fully unsupervised. The topics are "discovered" automatically from the data by trying to maximize the likelihood of observing the documents in your corpus, given the modeling assumptions. They are expected to capture some latent structure and organization within the documents, and often have a meaningful human interpretation for people familiar with the subject material.

We'll again turn to gensim to assist with data preparation and modeling. In particular, gensim offers a high-performance parallelized implementation of LDA with its [**LdaMulticore**](https://radimrehurek.com/gensim/models/ldamulticore.html) class.

In [102]:
from gensim.corpora import Dictionary, MmCorpus
from gensim.models.ldamulticore import LdaMulticore

import pyLDAvis
import pyLDAvis.gensim
import warnings
import pickle as pickle

The first step to creating an LDA model is to learn the full vocabulary of the corpus to be modeled. We'll use gensim's [**Dictionary**](https://radimrehurek.com/gensim/corpora/dictionary.html) class for this.

In [90]:
trigram_dictionary_filepath = os.path.join(intermediate_directory,
                                           'trigram_dict_all.dict')

In [91]:
%%time

# this is a bit time consuming - make the if statement True
# if you want to learn the dictionary yourself.
if 0 == 0:

    trigram_reviews = LineSentence(trigram_reviews_filepath)

    # learn the dictionary by iterating over all of the reviews
    trigram_dictionary = Dictionary(trigram_reviews)
    
    # filter tokens that are very rare or too common from
    # the dictionary (filter_extremes) and reassign integer ids (compactify)
    trigram_dictionary.filter_extremes(no_below=10, no_above=0.4)
    trigram_dictionary.compactify()

    trigram_dictionary.save(trigram_dictionary_filepath)
    
# load the finished dictionary from disk
trigram_dictionary = Dictionary.load(trigram_dictionary_filepath)

CPU times: user 292 ms, sys: 4 ms, total: 296 ms
Wall time: 287 ms


Like many NLP techniques, LDA uses a simplifying assumption known as the [*bag-of-words* model](https://en.wikipedia.org/wiki/Bag-of-words_model). In the bag-of-words model, a document is represented by the counts of distinct terms that occur within it. Additional information, such as word order, is discarded. 

Using the gensim Dictionary we learned to generate a bag-of-words representation for each review. The `trigram_bow_generator` function implements this. We'll save the resulting bag-of-words reviews as a matrix.

In the following code, "bag-of-words" is abbreviated as `bow`.

In [92]:
trigram_bow_filepath = os.path.join(intermediate_directory,
                                    'trigram_bow_corpus_all.mm')

In [93]:
def trigram_bow_generator(filepath):
    """
    generator function to read reviews from a file
    and yield a bag-of-words representation
    """
    
    for review in LineSentence(filepath):
        yield trigram_dictionary.doc2bow(review)

In [94]:
%%time

# this is a bit time consuming - make the if statement True
# if you want to build the bag-of-words corpus yourself.
if 0 == 0:

    # generate bag-of-words representations for
    # all reviews and save them as a matrix
    MmCorpus.serialize(trigram_bow_filepath,
                       trigram_bow_generator(trigram_reviews_filepath))
    
# load the finished bag-of-words corpus from disk
trigram_bow_corpus = MmCorpus(trigram_bow_filepath)

CPU times: user 324 ms, sys: 0 ns, total: 324 ms
Wall time: 327 ms


With the bag-of-words corpus, we're finally ready to learn our topic model from the reviews. We simply need to pass the bag-of-words matrix and Dictionary from our previous steps to `LdaMulticore` as inputs, along with the number of topics the model should learn. For this demo, we're asking for 50 topics.

In [95]:
lda_model_filepath = os.path.join(intermediate_directory, 'lda_model_all')

In [96]:
%%time

# this is a bit time consuming - make the if statement True
# if you want to train the LDA model yourself.
if 0 == 0:

    with warnings.catch_warnings():
        warnings.simplefilter('ignore')
        
        # workers => sets the parallelism, and should be
        # set to your number of physical cores minus one
        lda = LdaMulticore(trigram_bow_corpus,
                           num_topics=50,
                           id2word=trigram_dictionary,
                           workers=3)
    
    lda.save(lda_model_filepath)
    
# load the finished LDA model from disk
lda = LdaMulticore.load(lda_model_filepath)

CPU times: user 14.5 s, sys: 704 ms, total: 15.2 s
Wall time: 14.6 s


Our topic model is now trained and ready to use! Since each topic is represented as a mixture of tokens, you can manually inspect which tokens have been grouped together into which topics to try to understand the patterns the model has discovered in the data.

In [97]:
def explore_topic(topic_number, topn=25):
    """
    accept a user-supplied topic number and
    print out a formatted list of the top terms
    """
        
    print (u'{:20} {}'.format(u'term', u'frequency') + u'\n')

    for term, frequency in lda.show_topic(topic_number, topn=25):
        print (u'{:20} {:.3f}'.format(term, round(frequency, 3)))

In [98]:
explore_topic(topic_number=0)

term                 frequency

order                0.014
place                0.012
salad                0.011
price                0.010
service              0.010
friend               0.008
like                 0.008
restaurant           0.008
menu                 0.007
staff                0.007
try                  0.006
do_not               0.006
small                0.006
nice                 0.006
$                    0.006
soup                 0.005
quality              0.005
diner                0.004
thing                0.004
love                 0.004
sandwich             0.004
waitress             0.004
taste                0.004
und                  0.004
hot                  0.004


The first topic has strong associations with words like *taco*, *salsa*, *chip*, *burrito*, and *margarita*, as well as a handful of more general words. You might call this the **Mexican food** topic!

It's possible to go through and inspect each topic in the same way, and try to assign a human-interpretable label that captures the essence of each one. I've given it a shot for all 50 topics below.

In [99]:
topic_names = {0: u'mexican',
               1: u'menu',
               2: u'thai',
               3: u'steak',
               4: u'donuts & appetizers',
               5: u'specials',
               6: u'soup',
               7: u'wings, sports bar',
               8: u'foreign language',
               9: u'las vegas',
               10: u'chicken',
               11: u'aria buffet',
               12: u'noodles',
               13: u'ambience & seating',
               14: u'sushi',
               15: u'arizona',
               16: u'family',
               17: u'price',
               18: u'sweet',
               19: u'waiting',
               20: u'general',
               21: u'tapas',
               22: u'dirty',
               23: u'customer service',
               24: u'restrooms',
               25: u'chinese',
               26: u'gluten free',
               27: u'pizza',
               28: u'seafood',
               29: u'amazing',
               30: u'eat, like, know, want',
               31: u'bars',
               32: u'breakfast',
               33: u'location & time',
               34: u'italian',
               35: u'barbecue',
               36: u'arizona',
               37: u'indian',
               38: u'latin & cajun',
               39: u'burger & fries',
               40: u'vegetarian',
               41: u'lunch buffet',
               42: u'customer service',
               43: u'taco, ice cream',
               44: u'high cuisine',
               45: u'healthy',
               46: u'salad & sandwich',
               47: u'greek',
               48: u'poor experience',
               49: u'wine & dine'}

In [100]:
topic_names_filepath = os.path.join(intermediate_directory, 'topic_names.pkl')

with open(topic_names_filepath, 'w') as f:
    pickle.dump(topic_names, f)

TypeError: write() argument must be str, not bytes

You can see that, along with **mexican**, there are a variety of topics related to different styles of food, such as **thai**, **steak**, **sushi**, **pizza**, and so on. In addition, there are topics that are more related to the overall restaurant *experience*, like **ambience & seating**, **good service**, **waiting**, and **price**.

Beyond these two categories, there are still some topics that are difficult to apply a meaningful human interpretation to, such as topic 30 and 43.

Manually reviewing the top terms for each topic is a helpful exercise, but to get a deeper understanding of the topics and how they relate to each other, we need to visualize the data &mdash; preferably in an interactive format. Fortunately, we have the fantastic [**pyLDAvis**](https://pyldavis.readthedocs.io/en/latest/readme.html) library to help with that!

pyLDAvis includes a one-line function to take topic models created with gensim and prepare their data for visualization.

In [103]:
LDAvis_data_filepath = os.path.join(intermediate_directory, 'ldavis_prepared')

In [105]:
%%time

# this is a bit time consuming - make the if statement True
# if you want to execute data prep yourself.
if 0 == 0:

    LDAvis_prepared = pyLDAvis.gensim.prepare(lda, trigram_bow_corpus,
                                              trigram_dictionary)

#     with open(LDAvis_data_filepath, 'w') as f:
#         pickle.dump(LDAvis_prepared, f)
        
# # load the pre-prepared pyLDAvis data from disk
# with open(LDAvis_data_filepath) as f:
#     LDAvis_prepared = pickle.load(f)

/usr/local/lib/python3.5/dist-packages/pyLDAvis/_prepare.py:387: DeprecationWarning: 
.ix is deprecated. Please use
.loc for label based indexing or
.iloc for positional indexing

See the documentation here:
http://pandas.pydata.org/pandas-docs/stable/indexing.html#deprecate_ix
  topic_term_dists = topic_term_dists.ix[topic_order]


CPU times: user 7.64 s, sys: 144 ms, total: 7.78 s
Wall time: 9.06 s


`pyLDAvis.display(...)` displays the topic model visualization in-line in the notebook.

In [106]:
pyLDAvis.display(LDAvis_prepared)

### Wait, what am I looking at again?
There are a lot of moving parts in the visualization. Here's a brief summary:

* On the left, there is a plot of the "distance" between all of the topics (labeled as the _Intertopic Distance Map_)
  * The plot is rendered in two dimensions according a [*multidimensional scaling (MDS)*](https://en.wikipedia.org/wiki/Multidimensional_scaling) algorithm. Topics that are generally similar should be appear close together on the plot, while *dis*similar topics should appear far apart.
  * The relative size of a topic's circle in the plot corresponds to the relative frequency of the topic in the corpus.
  * An individual topic may be selected for closer scrutiny by clicking on its circle, or entering its number in the "selected topic" box in the upper-left.
* On the right, there is a bar chart showing top terms.
  * When no topic is selected in the plot on the left, the bar chart shows the top-30 most "salient" terms in the corpus. A term's *saliency* is a measure of both how frequent the term is in the corpus and how "distinctive" it is in distinguishing between different topics.
  * When a particular topic is selected, the bar chart changes to show the top-30 most "relevant" terms for the selected topic. The relevance metric is controlled by the parameter $\lambda$, which can be adjusted with a slider above the bar chart.
    * Setting the $\lambda$ parameter close to 1.0 (the default) will rank the terms solely according to their probability within the topic.
    * Setting $\lambda$ close to 0.0 will rank the terms solely according to their "distinctiveness" or "exclusivity" within the topic &mdash; i.e., terms that occur *only* in this topic, and do not occur in other topics.
    * Setting $\lambda$ to values between 0.0 and 1.0 will result in an intermediate ranking, weighting term probability and exclusivity accordingly.
* Rolling the mouse over a term in the bar chart on the right will cause the topic circles to resize in the plot on the left, to show the strength of the relationship between the topics and the selected term.

A more detailed explanation of the pyLDAvis visualization can be found [here](https://cran.r-project.org/web/packages/LDAvis/vignettes/details.pdf). Unfortunately, though the data used by gensim and pyLDAvis are the same, they don't use the same ID numbers for topics. If you need to match up topics in gensim's `LdaMulticore` object and pyLDAvis' visualization, you have to dig through the terms manually.

### Analyzing our LDA model
The interactive visualization pyLDAvis produces is helpful for both:
1. Better understanding and interpreting individual topics, and
1. Better understanding the relationships between the topics.

For (1), you can manually select each topic to view its top most freqeuent and/or "relevant" terms, using different values of the $\lambda$ parameter. This can help when you're trying to assign a human interpretable name or "meaning" to each topic.

For (2), exploring the _Intertopic Distance Plot_ can help you learn about how topics relate to each other, including potential higher-level structure between groups of topics.

In our plot, there is a stark divide along the x-axis, with two topics far to the left and most of the remaining 48 far to the right. Inspecting the two outlier topics provides a plausible explanation: both topics contain many non-English words, while most of the rest of the topics are in English. So, one of the main attributes that distinguish the reviews in the dataset from one another is their language.

This finding isn't entirely a surprise. In addition to English-speaking cities, the Yelp dataset includes reviews of businesses in Montreal and Karlsruhe, Germany, often written in French and German, respectively. Multiple languages isn't a problem for our demo, but for a real NLP application, you might need to ensure that the text you're processing is written in English (or is at least tagged for language) before passing it along to some downstream processing. If that were the case, the divide along the x-axis in the topic plot would immediately alert you to a potential data quality issue.

The y-axis separates two large groups of topics &mdash; let's call them "super-topics" &mdash; one in the upper-right quadrant and the other in the lower-right quadrant. These super-topics correlate reasonably well with the pattern we'd noticed while naming the topics:
* The super-topic in the *lower*-right tends to be about *food*. It groups together the **burger & fries**, **breakfast**, **sushi**, **barbecue**, and **greek** topics, among others.
* The super-topic in the *upper*-right tends to be about other elements of the *restaurant experience*. It groups together the **ambience & seating**, **location & time**, **family**, and **customer service** topics, among others.

So, in addition to the 50 direct topics the model has learned, our analysis suggests a higher-level pattern in the data. Restaurant reviewers in the Yelp dataset talk about two main things in their reviews, in general: (1) the food, and (2) their overall restaurant experience. For this dataset, this is a very intuitive result, and we probably didn't need a sophisticated modeling technique to tell it to us. When working with datasets from other domains, though, such high-level patterns may be much less obvious from the outset &mdash; and that's where topic modeling can help.

### Describing text with LDA
Beyond data exploration, one of the key uses for an LDA model is providing a compact, quantitative description of natural language text. Once an LDA model has been trained, it can be used to represent free text as a mixture of the topics the model learned from the original corpus. This mixture can be interpreted as a probability distribution across the topics, so the LDA representation of a paragraph of text might look like 50% _Topic A_, 20% _Topic B_, 20% _Topic C_, and 10% _Topic D_.

To use an LDA model to generate a vector representation of new text, you'll need to apply any text preprocessing steps you used on the model's training corpus to the new text, too. For our model, the preprocessing steps we used include:
1. Using spaCy to remove punctuation and lemmatize the text
1. Applying our first-order phrase model to join word pairs
1. Applying our second-order phrase model to join longer phrases
1. Removing stopwords
1. Creating a bag-of-words representation

Once you've applied these preprocessing steps to the new text, it's ready to pass directly to the model to create an LDA representation. The `lda_description(...)` function will perform all these steps for us, including printing the resulting topical description of the input text.

In [107]:
def get_sample_review(review_number):
    """
    retrieve a particular review index
    from the reviews file and return it
    """
    
    return list(it.islice(line_review(review_txt_filepath),
                          review_number, review_number+1))[0]

In [111]:
def lda_description(review_text, min_topic_freq=0.05):
    """
    accept the original text of a review and (1) parse it with spaCy,
    (2) apply text pre-proccessing steps, (3) create a bag-of-words
    representation, (4) create an LDA representation, and
    (5) print a sorted list of the top topics in the LDA representation
    """
    
    # parse the review text with spaCy
    parsed_review = nlp(review_text)
    
    # lemmatize the text and remove punctuation and whitespace
    unigram_review = [token.lemma_ for token in parsed_review
                      if not punct_space(token)]
    
    # apply the first-order and secord-order phrase models
    bigram_review = bigram_model[unigram_review]
    trigram_review = trigram_model[bigram_review]
    
    # remove any remaining stopwords
    trigram_review = [term for term in trigram_review
                      if not term in spacy.en.STOPWORDS]
    
    # create a bag-of-words representation
    review_bow = trigram_dictionary.doc2bow(trigram_review)
    
    # create an LDA representation
    review_lda = lda[review_bow]
    
    # sort with the most highly related topics first
    review_lda = sorted(review_lda, key=lambda(topic_number, freq): -freq)
    
    for topic_number, freq in review_lda:
        if freq < min_topic_freq:
            break
            
        # print the most highly related topic names and frequencies
        print ('{:25} {}'.format(topic_names[topic_number],
                                round(freq, 3)))

SyntaxError: invalid syntax (<ipython-input-111-fead63d164a4>, line 31)

In [109]:
sample_review = get_sample_review(50)
print (sample_review)

Mediocre food. 50$ for one person. I could have had better food at Panda Express. The orange peel chicken was cold and lacking flavor. The veggies in the dish tasted like they weren't washed before being cooked. Only good thing about this place is the atmosphere, it is the only reason I didn't give it 1 star.



In [110]:
lda_description(sample_review)

NameError: name 'lda_description' is not defined

In [112]:
sample_review = get_sample_review(100)
print (sample_review)

Went here for dinner on a weekday and the live music was great, makes the experience much more enjoyable.  The service was good and not too over-the-top.

We loved the curry chicken, the garlic naan was good too, but I prefer regular naan.  Their spicy shrimp was very flavourful.  The quality of the dishes was very high.

The portions are much smaller than what I am used to at other Indian places though and it's a bit pricey too.  Overall, it was great though, I'd come back.



In [113]:
lda_description(sample_review)

NameError: name 'lda_description' is not defined

## Word Vector Embedding with Word2Vec

Pop quiz! Can you complete this text snippet?

<br><br>

![word2vec quiz](https://s3.amazonaws.com/skipgram-images/word2vec-1.png)

<br><br><br>
You just demonstrated the core machine learning concept behind word vector embedding models!
<br><br><br>

![word2vec quiz 2](https://s3.amazonaws.com/skipgram-images/word2vec-2.png)

The goal of *word vector embedding models*, or *word vector models* for short, is to learn dense, numerical vector representations for each term in a corpus vocabulary. If the model is successful, the vectors it learns about each term should encode some information about the *meaning* or *concept* the term represents, and the relationship between it and other terms in the vocabulary. Word vector models are also fully unsupervised &mdash; they learn all of these meanings and relationships solely by analyzing the text of the corpus, without any advance knowledge provided.

Perhaps the best-known word vector model is [word2vec](https://arxiv.org/pdf/1301.3781v3.pdf), originally proposed in 2013. The general idea of word2vec is, for a given *focus word*, to use the *context* of the word &mdash; i.e., the other words immediately before and after it &mdash; to provide hints about what the focus word might mean. To do this, word2vec uses a *sliding window* technique, where it considers snippets of text only a few tokens long at a time.

At the start of the learning process, the model initializes random vectors for all terms in the corpus vocabulary. The model then slides the window across every snippet of text in the corpus, with each word taking turns as the focus word. Each time the model considers a new snippet, it tries to learn some information about the focus word based on the surrouding context, and it "nudges" the words' vector representations accordingly. One complete pass sliding the window across all of the corpus text is known as a training *epoch*. It's common to train a word2vec model for multiple passes/epochs over the corpus. Over time, the model rearranges the terms' vector representations such that terms that frequently appear in similar contexts have vector representations that are *close* to each other in vector space.

For a deeper dive into word2vec's machine learning process, see [here](https://arxiv.org/pdf/1411.2738v4.pdf).

Word2vec has a number of user-defined hyperparameters, including:
- The dimensionality of the vectors. Typical choices include a few dozen to several hundred.
- The width of the sliding window, in tokens. Five is a common default choice, but narrower and wider windows are possible.
- The number of training epochs.

For using word2vec in Python, [gensim](https://rare-technologies.com/deep-learning-with-word2vec-and-gensim/) comes to the rescue again! It offers a [highly-optimized](https://rare-technologies.com/word2vec-in-python-part-two-optimizing/), [parallelized](https://rare-technologies.com/parallelizing-word2vec-in-python/) implementation of the word2vec algorithm with its [Word2Vec](https://radimrehurek.com/gensim/models/word2vec.html) class.

In [114]:
from gensim.models import Word2Vec

trigram_sentences = LineSentence(trigram_sentences_filepath)
word2vec_filepath = os.path.join(intermediate_directory, 'word2vec_model_all')

We'll train our word2vec model using the normalized sentences with our phrase models applied. We'll use 100-dimensional vectors, and set up our training process to run for twelve epochs.

In [115]:
%%time

# this is a bit time consuming - make the if statement True
# if you want to train the word2vec model yourself.
if 0 == 0:

    # initiate the model and perform the first epoch of training
    food2vec = Word2Vec(trigram_sentences, size=100, window=5,
                        min_count=20, sg=1, workers=4)
    
    food2vec.save(word2vec_filepath)

    # perform another 11 epochs of training
    for i in range(1,12):

        food2vec.train(trigram_sentences)
        food2vec.save(word2vec_filepath)
        
# load the finished model from disk
food2vec = Word2Vec.load(word2vec_filepath)
food2vec.init_sims()

print (u'{} training epochs so far.'.format(food2vec.train_count))

ValueError: You must specify either total_examples or total_words, for proper alpha and progress calculations. The usual value is total_examples=model.corpus_count.

On my four-core machine, each epoch over all the text in the ~1 million Yelp reviews takes about 5-10 minutes.

In [60]:
print (u'{:,} terms in the food2vec vocabulary.'.format(len(food2vec.vocab)))

50,835 terms in the food2vec vocabulary.


Let's take a peek at the word vectors our model has learned. We'll create a pandas DataFrame with the terms as the row labels, and the 100 dimensions of the word vector model as the columns.

In [90]:
# build a list of the terms, integer indices,
# and term counts from the food2vec model vocabulary
ordered_vocab = [(term, voc.index, voc.count)
                 for term, voc in food2vec.vocab.iteritems()]

# sort by the term counts, so the most common terms appear first
ordered_vocab = sorted(ordered_vocab, key=lambda (term, index, count): -count)

# unzip the terms, integer indices, and counts into separate lists
ordered_terms, term_indices, term_counts = zip(*ordered_vocab)

# create a DataFrame with the food2vec vectors as data,
# and the terms as row labels
word_vectors = pd.DataFrame(food2vec.syn0norm[term_indices, :],
                            index=ordered_terms)

word_vectors

0         1         2         3         4   \
the                      -0.035762 -0.173890 -0.035782 -0.007144  0.032371   
be                       -0.074780 -0.049524  0.085974 -0.098892  0.141556   
and                      -0.070505 -0.026918  0.028344 -0.099909  0.127974   
i                        -0.161238  0.050831 -0.081706 -0.084479  0.053073   
a                        -0.083491 -0.033712 -0.124125 -0.110776 -0.033046   
to                       -0.012082  0.033135 -0.063183 -0.057252 -0.018721   
it                        0.025022  0.081581  0.127987 -0.188015  0.041450   
have                     -0.140812 -0.070552  0.022102  0.001077  0.109890   
of                       -0.036341 -0.054903  0.000644 -0.010602  0.168195   
not                      -0.075276  0.109047  0.055135  0.052251  0.209437   
for                      -0.102976  0.001369 -0.069402 -0.122936  0.028278   
in                       -0.053390 -0.175599 -0.091688 -0.153791  0.003205   
we                       -0.015929 -0.019187 -0.186680 -0.240963  0.077926   
that                     -0.026609  0.085940  0.118164  0.011576  0.156952   
but                      -0.063436  0.089140 -0.057425 -0.093110  0.066531   
with                      0.042318 -0.186670 -0.230563  0.076302  0.216593   
my                       -0.002067 -0.023159  0.035879  0.036316 -0.110738   
this                      0.139739  0.184600  0.137359 -0.109916  0.021484   
you                      -0.171262 -0.119866  0.063801 -0.087287 -0.061923   
on                        0.057481  0.044937 -0.063766 -0.007839  0.161119   
they                     -0.235321 -0.026314  0.143165 -0.170460  0.042189   
food                     -0.164133  0.007745  0.058311 -0.169839 -0.042278   
do                        0.091428 -0.132115  0.105080  0.135949  0.038100   
good                     -0.239592 -0.232940 -0.005036 -0.028226  0.149816   
place                     0.025479  0.130311  0.119834 -0.096365  0.013793   
so                        0.021455  0.079794  0.192058 -0.093809 -0.094279   
get                       0.009313 -0.101684 -0.163864 -0.159002  0.018936   
go                        0.031094 -0.126839 -0.054429 -0.221885 -0.063464   
at                        0.102501 -0.095756 -0.216304 -0.107230 -0.112544   
as                        0.141091  0.073669  0.109637 -0.112564 -0.167600   
...                            ...       ...       ...       ...       ...   
hard_boiled_egg           0.025154  0.060949 -0.064816  0.071975  0.087870   
poached_quail_egg         0.060616  0.001519 -0.052908  0.013590  0.031732   
egg_foo                   0.022816  0.077288 -0.122898 -0.022765 -0.137979   
sherri                   -0.089536 -0.030370  0.011311 -0.094766 -0.121970   
hindrance                -0.033325  0.011520  0.027370  0.240223 -0.004098   
eggdrop_soup             -0.028369  0.040039 -0.005378 -0.088736  0.015203   
arbitrarily               0.203714 -0.047405 -0.045261  0.000302 -0.074105   
faisant                   0.046053  0.083873  0.057943  0.174203 -0.121259   
marian                   -0.035330  0.146843 -0.173594 -0.010971 -0.150150   
9:30p                     0.119010  0.002159 -0.000148 -0.102635 -0.016918   
extra_chasu              -0.225889 -0.131615  0.046431  0.017999  0.119188   
lavosh_wrap               0.134420 -0.032055  0.012240  0.024420  0.031334   
dum_dum                  -0.002741 -0.137371  0.030704 -0.030365 -0.134645   
triplet                  -0.010495  0.057432 -0.019535 -0.044881  0.042409   
nantucket                -0.124356  0.141918 -0.038579  0.035650 -0.157662   
gurl                     -0.119794  0.025898 -0.070130 -0.027929  0.113244   
nordstroms               -0.030410 -0.026861 -0.016836  0.097363 -0.098189   
eau_de                   -0.146450  0.000788 -0.094391  0.146833 -0.051392   
extra_.5                 -0.124972 -0.032596 -0.017800  0.106415 -0.086728   
hazelnut_crunch           0.055723 -0.098708  0.013225

Holy wall of numbers! This DataFrame has 50,835 rows &mdash; one for each term in the vocabulary &mdash; and 100 colums. Our model has learned a quantitative vector representation for each term, as expected.

Put another way, our model has "embedded" the terms into a 100-dimensional vector space.

### So... what can we do with all these numbers?
The first thing we can use them for is to simply look up related words and phrases for a given term of interest.

In [63]:
def get_related_terms(token, topn=10):
    """
    look up the topn most similar terms to token
    and print them as a formatted list
    """

    for word, similarity in food2vec.most_similar(positive=[token], topn=topn):

        print (u'{:20} {}'.format(word, round(similarity, 3)))

### What things are like Burger King?

In [64]:
get_related_terms(u'burger_king')

mcdonalds            0.895
wendy_'s             0.855
mcd_'s               0.853
mcdonald_'s          0.852
denny_'s             0.816
bk                   0.808
carl_'s_jr.          0.8
red_robin            0.792
mickey_d_'s          0.771
sonic                0.765


The model has learned that fast food restaurants are similar to each other! In particular, *mcdonalds* and *wendy's* are the most similar to Burger King, according to this dataset. In addition, the model has found that alternate spellings for the same entities are probably related, such as *mcdonalds*, *mcdonald's* and *mcd's*.

### When is happy hour?

In [65]:
get_related_terms(u'happy_hour')

hh                   0.874
reverse_happy_hour   0.801
happy_hr             0.771
during_happy_hour    0.672
mon_fri              0.634
3_6pm                0.632
hh.                  0.631
4_7pm                0.625
special              0.621
happy_hour_3_6pm     0.618


The model has noticed several alternate spellings for happy hour, such as *hh* and *happy hr*, and assesses them as highly related. If you were looking for reviews about happy hour, such alternate spellings would be very helpful to know.

Taking a deeper look &mdash; the model has turned up phrases like *3-6pm*, *4-7pm*, and *mon-fri*, too. This is especially interesting, because the model has no advance knowledge at all about what happy hour is, and what time of day it should be. But simply by scanning through restaurant reviews, the model has discovered that the concept of happy hour has something very important to do with that block of time around 3-7pm on weekdays.

### Let's make pasta tonight. Which style do you want?

In [66]:
get_related_terms(u'pasta', topn=20)

lasagna              0.798
spaghetti            0.773
bolognese            0.757
fettuccine           0.748
penne                0.745
rigatoni             0.743
angel_hair           0.721
linguine             0.716
angel_hair_pasta     0.712
penne_pasta          0.712
carbonara            0.705
fettucini            0.704
alfredo              0.703
tortellini           0.703
manicotti            0.7
ziti                 0.698
gnocci               0.694
ravioli              0.694
linguini             0.693
risotto              0.691


## Word algebra!
No self-respecting word2vec demo would be complete without a healthy dose of *word algebra*, also known as *analogy completion*.

The core idea is that once words are represented as numerical vectors, you can do math with them. The mathematical procedure goes like this:
1. Provide a set of words or phrases that you'd like to add or subtract.
1. Look up the vectors that represent those terms in the word vector model.
1. Add and subtract those vectors to produce a new, combined vector.
1. Look up the most similar vector(s) to this new, combined vector via cosine similarity.
1. Return the word(s) associated with the similar vector(s).

But more generally, you can think of the vectors that represent each word as encoding some information about the *meaning* or *concepts* of the word. What happens when you ask the model to combine the meaning and concepts of words in new ways? Let's see.

In [67]:
def word_algebra(add=[], subtract=[], topn=1):
    """
    combine the vectors associated with the words provided
    in add= and subtract=, look up the topn most similar
    terms to the combined vector, and print the result(s)
    """
    answers = food2vec.most_similar(positive=add, negative=subtract, topn=topn)
    
    for term, similarity in answers:
        print (term)

### breakfast + lunch = ?
Let's start with a softball.

In [68]:
word_algebra(add=[u'breakfast', u'lunch'])

brunch


OK, so the model knows that *brunch* is a combination of *breakfast* and *lunch*. What else?

### lunch - day + night = ?

In [69]:
word_algebra(add=[u'lunch', u'night'], subtract=[u'day'])

dinner


Now we're getting a bit more nuanced. The model has discovered that:
- Both *lunch* and *dinner* are meals
- The main difference between them is time of day
- Day and night are times of day
- Lunch is associated with day, and dinner is associated with night

What else?

### taco - mexican + chinese = ?

In [70]:
word_algebra(add=[u'taco', u'chinese'], subtract=[u'mexican'])

dumpling


Here's an entirely new and different type of relationship that the model has learned.
- It knows that tacos are a characteristic example of Mexican food
- It knows that Mexican and Chinese are both styles of food
- If you subtract *Mexican* from *taco*, you're left with something like the concept of a *"characteristic type of food"*, which is represented as a new vector
- If you add that new *"characteristic type of food"* vector to Chinese, you get *dumpling*.

What else?

### bun - american + mexican = ?

In [71]:
word_algebra(add=[u'bun', u'mexican'], subtract=[u'american'])

tortilla


The model knows that both *buns* and *tortillas* are the doughy thing that goes on the outside of your real food, and that the primary difference between them is the style of food they're associated with.

What else?

### filet mignon - beef + seafood = ?

In [72]:
word_algebra(add=[u'filet_mignon', u'seafood'], subtract=[u'beef'])

raw_oyster


The model has learned a concept of *delicacy*. If you take filet mignon and subtract beef from it, you're left with a vector that roughly corresponds to delicacy. If you add the delicacy vector to *seafood*, you get *raw oyster*.

What else?

### coffee - drink + snack = ?

In [73]:
word_algebra(add=[u'coffee', u'snack'], subtract=[u'drink'])

pastry


The model knows that if you're on your coffee break, but instead of drinking something, you're eating something... that thing is most likely a pastry.

What else?

### Burger King + fine dining = ?

In [74]:
word_algebra(add=[u'burger_king', u'fine_dining'])

denny_'s


Touché. It makes sense, though. The model has learned that both Burger King and Denny's are large chains, and that both serve fast, casual, American-style food. But Denny's has some elements that are slightly more upscale, such as printed menus and table service. Fine dining, indeed.

*What if we keep going?*

### Denny's + fine dining = ?

In [75]:
word_algebra(add=[u"denny_'s", u'fine_dining'])

applebee_'s


This seems like a good place to land... what if we explore the vector space around *Applebee's* a bit, in a few different directions? Let's see what we find.

#### Applebee's + italian = ?

In [76]:
word_algebra(add=[u"applebee_'s", u'italian'])

olive_garden


#### Applebee's + pancakes = ?

In [77]:
word_algebra(add=[u"applebee_'s", u'pancakes'])

ihop


#### Applebee's + pizza = ?

In [78]:
word_algebra(add=[u"applebee_'s", u'pizza'])

pizza_hut


You could do this all day. One last analogy before we move on...

### wine - grapes + barley = ?

In [79]:
word_algebra(add=[u'wine', u'barley'], subtract=[u'grapes'])

beer


## Word Vector Visualization with t-SNE

[t-Distributed Stochastic Neighbor Embedding](https://lvdmaaten.github.io/publications/papers/JMLR_2008.pdf), or *t-SNE* for short, is a dimensionality reduction technique to assist with visualizing high-dimensional datasets. It attempts to map high-dimensional data onto a low two- or three-dimensional representation such that the relative distances between points are preserved as closely as possible in both high-dimensional and low-dimensional space.

scikit-learn provides a convenient implementation of the t-SNE algorithm with its [TSNE](http://scikit-learn.org/stable/modules/generated/sklearn.manifold.TSNE.html) class.

In [80]:
from sklearn.manifold import TSNE

Our input for t-SNE will be the DataFrame of word vectors we created before. Let's first:
1. Drop stopwords &mdash; it's probably not too interesting to visualize *the*, *of*, *or*, and so on
1. Take only the 5,000 most frequent terms in the vocabulary &mdash; no need to visualize all ~50,000 terms right now.

In [81]:
tsne_input = word_vectors.drop(spacy.en.STOPWORDS, errors=u'ignore')
tsne_input = tsne_input.head(5000)

In [82]:
tsne_input.head()

0         1         2         3         4         5         6   \
food  -0.164133  0.007745  0.058311 -0.169839 -0.042278  0.004095  0.203732   
good  -0.239592 -0.232940 -0.005036 -0.028226  0.149816 -0.133312 -0.034164   
place  0.025479  0.130311  0.119834 -0.096365  0.013793  0.074431 -0.063780   
order  0.045996 -0.035101 -0.045906 -0.280336  0.157393 -0.146304 -0.064311   
great -0.189524 -0.259744 -0.041182  0.021664  0.132266 -0.030005 -0.078055   

             7         8         9     ...           90        91        92  \
food  -0.021252 -0.084491 -0.016372    ...    -0.096869  0.060159 -0.133541   
good  -0.130310 -0.013757  0.008618    ...    -0.034984 -0.135347 -0.112965   
place  0.063191 -0.004273  0.111458    ...    -0.076251 -0.076574 -0.086146   
order -0.082754  0.124693 -0.194072    ...    -0.063306  0.125928 -0.194433   
great -0.004111 -0.016184  0.054305    ...    -0.052275 -0.122914 -0.075555   

             93        94        95        96        97        98        99  
food   0.166804  0.084901  0.109261  0.137871  0.018093 -0.158754 -0.042917  
good   0.056312  0.055106 -0.026181 -0.135510  0.087664  0.009934 -0.111619  
place -0.023936  0.136419 -0.001543 -0.084301  0.016356 -0.148379 -0.016498  
order -0.129551  0.039680  0.058868 -0.023189 -0.153715  0.152482 -0.003842  
great  0.193383  0.017008 -0.117156 -0.041065  0.140430 -0.067349 -0.040291  

[5 rows x 100 columns]

In [83]:
tsne_filepath = os.path.join(intermediate_directory,
                             u'tsne_model')

tsne_vectors_filepath = os.path.join(intermediate_directory,
                                     u'tsne_vectors.npy')

In [93]:
%%time

if 0 == 1:
    
    tsne = TSNE()
    tsne_vectors = tsne.fit_transform(tsne_input.values)
    
    with open(tsne_filepath, 'w') as f:
        pickle.dump(tsne, f)

    pd.np.save(tsne_vectors_filepath, tsne_vectors)
    
with open(tsne_filepath) as f:
    tsne = pickle.load(f)
    
tsne_vectors = pd.np.load(tsne_vectors_filepath)

tsne_vectors = pd.DataFrame(tsne_vectors,
                            index=pd.Index(tsne_input.index),
                            columns=[u'x_coord', u'y_coord'])

CPU times: user 31.3 ms, sys: 2.47 ms, total: 33.8 ms
Wall time: 32.9 ms


Now we have a two-dimensional representation of our data! Let's take a look.

In [94]:
tsne_vectors.head()

x_coord   y_coord
food   2.313886  6.475995
good   8.763030  4.633407
place -8.942178  2.221976
order -2.876029 -2.300830
great  9.515772  5.076319

In [95]:
tsne_vectors[u'word'] = tsne_vectors.index

### Plotting with Bokeh

In [88]:
from bokeh.plotting import figure, show, output_notebook
from bokeh.models import HoverTool, ColumnDataSource, value

output_notebook()

Loading BokehJS ...

In [89]:
# add our DataFrame as a ColumnDataSource for Bokeh
plot_data = ColumnDataSource(tsne_vectors)

# create the plot and configure the
# title, dimensions, and tools
tsne_plot = figure(title=u't-SNE Word Embeddings',
                   plot_width = 800,
                   plot_height = 800,
                   tools= (u'pan, wheel_zoom, box_zoom,'
                           u'box_select, resize, reset'),
                   active_scroll=u'wheel_zoom')

# add a hover tool to display words on roll-over
tsne_plot.add_tools( HoverTool(tooltips = u'@word') )

# draw the words as circles on the plot
tsne_plot.circle(u'x_coord', u'y_coord', source=plot_data,
                 color=u'blue', line_alpha=0.2, fill_alpha=0.1,
                 size=10, hover_line_color=u'black')

# configure visual elements of the plot
tsne_plot.title.text_font_size = value(u'16pt')
tsne_plot.xaxis.visible = False
tsne_plot.yaxis.visible = False
tsne_plot.grid.grid_line_color = None
tsne_plot.outline_line_color = None

# engage!
show(tsne_plot);

## Conclusion

Whew! Let's round up the major components that we've seen:
1. Text processing with **spaCy**
1. Automated **phrase modeling**
1. Topic modeling with **LDA** $\ \longrightarrow\ $ visualization with **pyLDAvis**
1. Word vector modeling with **word2vec** $\ \longrightarrow\ $ visualization with **t-SNE**

#### Why use these models?
Dense vector representations for text like LDA and word2vec can greatly improve performance for a number of common, text-heavy problems like:
- Text classification
- Search
- Recommendations
- Question answering

...and more generally are a powerful way machines can help humans make sense of what's in a giant pile of text. They're also often useful as a pre-processing step for many other downstream machine learning applications.

## Data Science @ S&P Global &mdash; *we are hiring!*